serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,472 @@
+        temp = torch.tensor(
+            list_temperatures,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        top_ps = torch.tensor(
+            list_top_ps,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        top_ks = torch.tensor(
+            list_top_ks,
+            dtype=torch.int,
+            device="cpu",
+            pin_memory=True,
+        )
+        frequency_penalties = torch.tensor(
+            list_frequency_penalties,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        presence_penalties = torch.tensor(
+            list_presence_penalties,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        repetition_penalties = torch.tensor(
+            list_repetition_penalties,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        logit_bias_indices = torch.tensor(
+            list_logit_bias_indices,
+            dtype=torch.long,
+            device="cpu",
+            pin_memory=True,
+        )
+        # Convert 1-based index to 0-based
+        logit_bias_indices -= 1
+        logit_bias_values = torch.tensor(
+            list_logit_bias_values,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        past_output_tokens = torch.tensor(
+            list_past_output_tokens,
+            dtype=torch.long,
+            device="cpu",
+            pin_memory=True,
+        )
+
+        return cls(
+            mask_random,
+            mask_greedy,
+            mask_top_logprob,
+            temp.to(device=dev, non_blocking=True),
+            top_ps.to(device=dev, non_blocking=True),
+            top_ks.to(device=dev, non_blocking=True),
+            frequency_penalties.to(device=dev, non_blocking=True),
+            presence_penalties.to(device=dev, non_blocking=True),
+            repetition_penalties.to(device=dev, non_blocking=True),
+            logit_bias_indices.to(device=dev, non_blocking=True),
+            logit_bias_values.to(device=dev, non_blocking=True),
+            past_output_tokens.to(device=dev, non_blocking=True),
+        )
+
+
+@dataclass
+class SamplingMetadata:
+    # mapping for <batch index, tuple(sampling type, greedy/random sampling index)>
+    index_map: dict[int, Tuple[SamplingType, int]]
+    has_random: bool
+    has_greedy: bool
+    apply_top_p_top_k: bool
+    apply_penalty: bool
+    apply_bias: bool
+    # mapping for <batch index, Tuple(logprob, top-k, top-k index)>
+    logprob_index_map: dict[int, Tuple[bool, int, int]]
+    has_logprob: bool
+    sampling_tensors: SamplingTensors
+    sampling_params: List[SamplingParams]
+
+    @classmethod
+    def from_sampling_params(
+        cls,
+        sampling_params: List[SamplingParams],
```

Comment:
@sunggg can we find time to talk about refactoring `SamplingParams` both at `oLLM` and `mlc` layers?